[READY] Feature/update annotation position on file changes

thilo-behnke opened the pull request:
### Checklist
- [x] I run `yarn run webpack:build:main`: the project builds without errors.
- [x] I run `yarn lint`: there project builds without code style warnings.
- [x] I removed unnecessary whitespace changes in all related files.
- [x] I updated the documentation and models.
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] ~I added (end-to-end) test cases for the new functionality.~

### Motivation and Context
Annotations should be moved accordingly when the text is edited in the editor.
Otherwise they get displayed at the wrong line when the selected file is switched or the page is reloaded.

### Description
- subscribe to text changes to update the position of the error annotations
- persist the error annotation updates in a session ~~file, that is part of the assignment repository (session.json)~~
=> stored in localstorage
- load this session file on opening the programming exercise editor screen to restore the annotation positions
- check, whether the buildLogs or the session are newer - therefore a timestamp is used 

**Out of scope for this PR, could be done in follow-up tasks:**
- Optimization: The annotation update callback is triggered on every changed character. I don't see any performance issues atm, but it would be be good to buffer changes for a period of time and then process them all at once.
- Improvement: Also show an error symbol next to the files in the file browser, I made sure the code structure is already prepared for this feature.
- Improvement: Change the error icons to warning icons once the file is touched, because the errors might have been be resolved by the students changes
- Improvement: Additionally show markers at the position of the expression that caused the error. 

### Steps for Testing
1) Open a programming exercise in the editor
2) Make an error that is caught by the compliler (e.g. publi instead of public) and commit&run
3) When the errors show up after the build, remove/add line breaks and wait for the file saving to be finished
  -> If you switch between files, the error annotations stay in the correct lines
  -> If you reload the page, the annotations stay in the correct lines

### Screenshots
![error-editor](https://user-images.githubusercontent.com/28230611/54479233-04998e80-481b-11e9-88e5-b9a4a37d8611.jpg)

krusche commented:
Sorry, I probably merged your other pull request too early :-) 
I actually refactored the name to be code-editor instead of being simply editor in the develop branch. Please make sure to apply these changes soon to this branch as they lead to merge conflicts

thilo-behnke commented:
No problem, thanks for letting me know :)

krusche commented:
I do not really like the solution with the session.json file, because it will be in the repository of the student and we actually want to move away storing information in the student repository. One of the next tasks will be to store the README.md as problem statement of the programming exercise instead of storing it in the repo. Also at the moment, the online editor is quite performance intense, because it uses quite a lot REST calls. I would rather switch to web sockets here.

I suggest that we have a quick call next week and brainstorm how we can solve the problem of outdated annotations. Keep in mind that we do not provide a full IDE. Maybe it is fine, if annotations outdate in some cases and we only support updating them if the user handles them within the same browser session.

thilo-behnke commented on src/main/webapp/app/core/websocket/websocket.service.ts:
```diff
@@ -167,7 +167,14 @@ export class JhiWebsocketService implements OnDestroy {
             this.subscribers[channel] = this.stompClient.subscribe(channel, data => {
-                this.listenerObservers[channel].next(JSON.parse(data.body));
+                const res = JSON.parse(data.body);
+                if (!res.error) {
```
I think it's a good idea to do the check here, not in the components.
This way we can use subscribe(res => handleRes(), err => handleError()).
Otherwise we would have to do subscribe (res => !res.error? handleRes() : handleError()).

thilo-behnke commented:
- I migrated the update file call to a websocket
- I changed the session persistance to localStorage

Notes:
- There was no error handling implemented yet for a websocket, so I had to create something new. I also adjusted the websocket service in the frontend to have streamlined error handling.
- We should create a new PR if we want to migrate more REST calls
- There is no good solution for annotations if the user switches computers/browsers. The problem is that we can't differentiate a reload of the page and someone logging in from another computer/browser, so after every reload given that the git repo is not clean (uncommited changes), the annotations would not show.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/websocket/RepositoryWebsocketService.java:
```diff
@@ -0,0 +1,177 @@
+import java.security.Principal;
+import java.util.Optional;
+
+@Entity
```
do we really need the annotations @Entity and @Id here? These classes are not persisted in the database, so I guess we can remove them.
I would also suggest to put them into their own files as this is best practice in Java

thilo-behnke replied:
You are right about both - I made a commit to fix these issues.

krusche commented on src/main/webapp/app/code-editor/code-editor.component.ts:
```diff
@@ -151,18 +155,27 @@ export class CodeEditorComponent implements OnInit, OnChanges, OnDestroy {
-                text: safeUnescape(text)
-            }))
-            .reduce((acc, { fileName, ...rest }) => ({ ...acc, [fileName]: [...(acc[fileName] || []), rest] }), {});
+        const ts = buildLogs.length ? Date.parse(buildLogs[0].time) : 0;
```
I guess ts stands for timestamp, right? I would prefer if you could write timestamp instead to make it easier to read it

thilo-behnke replied:
True - I'll refactor this.

thilo-behnke replied:
Refactored.

krusche commented on src/main/webapp/app/code-editor/code-editor.component.ts:
```diff
@@ -151,18 +155,27 @@ export class CodeEditorComponent implements OnInit, OnChanges, OnDestroy {
+                    text: safeUnescape(text),
+                    ts: Date.parse(time)
+                }))
+                .reduce((acc, { fileName, ...rest }) => ({ ...acc, [fileName]: [...(acc[fileName] || []), rest] }), {});
```
what does acc stands for?

thilo-behnke replied:
It's short for accumulator, this term is also in used in official documentation and an informal javascript convention (e.g. https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce). There are other conventions as "prev", but none of them make much sense to me.

krusche replied:
I think accumulator is too generic. Can we find a better term here? If not, then please put in the full name 'accumulator' otherwise someone might think acc stands for account or something else.

Also add a comment what you want to achieve with reduce ;-) Otherwise the code is hard to understand. I could not tell from looking at it, what exactly you are doing here

thilo-behnke replied:
I'll change it and add some comments.

thilo-behnke replied:
Done.

krusche commented on src/main/webapp/app/code-editor/code-editor.component.ts:
```diff
@@ -197,6 +210,21 @@ export class CodeEditorComponent implements OnInit, OnChanges, OnDestroy {
+        if (this.participation) {
+            const sessions = JSON.parse(this.localStorageService.retrieve('sessions') || '{}');
+            this.session = sessions[this.participation.id];
+            if (this.session && (!this.buildLogErrors || this.session.ts > this.buildLogErrors.ts)) {
```
ts -> timestamp

thilo-behnke replied:
True - I'll refactor this.

thilo-behnke replied:
Done.

krusche commented on src/main/webapp/app/entities/ace-editor/annotation.model.ts:
```diff
@@ -1 +1 @@
-export type AceAnnotation = { row: string; column: string; text: string; type: string };
+export type AceAnnotation = { row: number; column: number; text: string; type: string; ts: number };
```
would it make sense to use a proper class type here?

thilo-behnke replied:
Actually thinking about it, it would make sense to move the updateAnnotation from the component to a proper Annotation class for reuse purposes. Good idea!

thilo-behnke replied:
Done.

krusche commented on src/main/webapp/app/entities/ace-editor/session.model.ts:
```diff
@@ -0,0 +1,3 @@
+import { AceAnnotation } from './annotation.model';
```
would it make sense to use a proper class type here?

krusche commented on src/main/webapp/app/entities/ace-editor/session.model.ts:
```diff
@@ -0,0 +1,3 @@
+import { AceAnnotation } from './annotation.model';
+
+export type Session = { ts: number; errors: { [fileName: string]: AceAnnotation[] } };
```
would it make sense to use a proper class type here?

thilo-behnke replied:
I think in this case Session is really just a plain object without any functionality attached to it. So I would suggest to keep it lightweight in this case.

krusche commented on src/main/webapp/app/entities/ace-editor/text-change.model.ts:
```diff
@@ -0,0 +1 @@
+export type TextChange = { action: string; start: { row: number; column: number }; end: { row: number; column: number } };
```
would it make sense to use a proper class type here?

thilo-behnke replied:
I think in this case TextChange is really just a plain object without any functionality attached to it. So I would suggest to keep it lightweight in this case.

krusche commented on src/main/webapp/app/code-editor/ace/code-editor-ace.component.ts:
```diff
@@ -81,14 +112,118 @@ export class CodeEditorAceComponent implements OnInit, AfterViewInit, OnChanges
      * @param {SimpleChanges} changes
```
Could you extend the documentation here and describe what kind of changes you do with the annotations?

thilo-behnke replied:
I will add a more explanatory comment to this section.

thilo-behnke replied:
Done.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/websocket/repository/FileSubmissionError.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.io.Serializable;
+
+public class FileSubmissionError extends WebsocketError implements Serializable {
```
JavaDoc missing

thilo-behnke replied:
JDoc added.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/websocket/repository/FileSubmissionSuccess.java:
```diff
@@ -0,0 +1,16 @@
+package de.tum.in.www1.artemis.web.websocket.repository;
+
+public class FileSubmissionSuccess {
```
JavaDoc missing

thilo-behnke replied:
JDoc added.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/websocket/repository/RepositoryWebsocketService.java:
```diff
@@ -0,0 +1,119 @@
+package de.tum.in.www1.artemis.web.websocket.repository;
+
+import afu.org.checkerframework.checker.nullness.qual.Nullable;
```
I guess you mean import javax.annotation.Nullable;
```suggestion
import javax.annotation.Nullable;
```

thilo-behnke replied:
You are right, thanks!

krusche commented on src/main/java/de/tum/in/www1/artemis/web/websocket/repository/RepositoryWebsocketService.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * PUT /repository/{participationId}/file: Update the file content
```
The JavaDoc comment could be a bit more descriptive. I am not sure if we should write 'PUT' as this is not a REST call anymore --> this is probably due to copy and paste

thilo-behnke replied:
Done.

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/websocket/repository/FileSubmission.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.io.Serializable;
+
+public class FileSubmission implements Serializable {
```
JDoc added.

thilo-behnke commented:
@krusche The requested changes are commited.
